packages/shell-api/src/shell-internal-state.ts

```diff
@@ -0,0 +1,238 @@
+      attr: [
+        {
+          name: 'use',
+          description: 'shell-api.help.help.use'
+        },
+        {
+          name: 'it',
+          description: 'shell-api.help.help.it'
+        },
+        {
+          name: 'show databases',
+          description: 'shell-api.help.help.show-databases'
+        },
+        {
+          name: 'show collections',
+          description: 'shell-api.help.help.show-collections'
+        },
+        {
+          name: '.exit',
+          description: 'shell-api.help.help.exit'
+        }
+      ]
+    });
+  }
+
+  /**
+   * Prepare a `contextObject` as global context and set it as context
+   * Add each attribute to the AsyncWriter also.
+   *
+   * The `contextObject` is prepared so that it can be used as global object
+   * for the repl evaluationi.
+   *
+   * @note The `contextObject` is mutated, it will retain all of its existing
+   * properties but also have the global shell api objects and functions.
+   *
+   * @param {Object} contextObject - contextObject an object used as global context.
+   */
+  setCtx(contextObject: any): void {
+    // Add API methods for VSCode and scripts
+    contextObject.use = this.use.bind(this);
+    contextObject.show = this.show.bind(this);
+    contextObject.it = this.it.bind(this);
+    contextObject.help = this.help.bind(this);
+    contextObject.toIterator = toIterator;
+    contextObject.print = async(arg) => {
+      if (arg.toReplString) {
+        console.log(await arg.toReplString());
+      } else {
+        console.log(arg);
+      }
+    };
+    contextObject.printjson = contextObject.print;
+    Object.assign(contextObject, ShellBson);
+    contextObject.rs = new ReplicaSet(this.currentDb.mongo);
+    contextObject.sh = new Shard(this.currentDb.mongo);
+
+    // Add global shell objects
+    this.asyncWriter.symbols.initializeApiObjects({
+      db: signatures.Database,
+      rs: signatures.ReplicaSet,
+      sh: signatures.Shard,
+      Mongo: signatures.Mongo
+    });
+
+    try {
+      Object.defineProperty(contextObject, 'db', {
+        set: (newDb: any) => {
+          if (newDb.shellApiType && newDb.shellApiType() !== 'Database') {
+            throw new MongoshInvalidInputError('Cannot reassign \'db\' to non-Database type');
+          }
+          this.currentDb = newDb;
+          contextObject.rs = new ReplicaSet(this.currentDb.mongo);
+          contextObject.sh = new Shard(this.currentDb.mongo);
+          this.fetchConnectionInfo();
+          return newDb;
+        },
+        get: () => {
+          return this.currentDb;
+        }
+      });
+    } catch (e) { // java shell, can't use getters/setters
+      console.log(`in java shell, platform: ${this.initialServiceProvider.platform}`);
+      contextObject.db = this.currentDb;
+    }
+
+    contextObject.Mongo = async(uri?, options?): Promise<Mongo> => {
+      if (
+        this.initialServiceProvider.platform !== ReplPlatform.CLI
```

Comment:
I love the idea to be able to negotiate with the service providers! I suggest had-hoc feature flags though, usually more helpful. Something like this:

```
if (!this.initialServiceProvider.features.multipleConnection) {
  // ...
}
```

or even a more dynamic one:

```
!this.initialServiceProvider.featureEnabled('multipleConnection')
```

Reply:
I have been thinking about this quite a bit, and generally agree with you about using ad-hoc flags but I wonder what that would look like when we are also creating such an extensive set of feature flags based on the platform enum in shell-api. So every function/class/etc in the Shell API has a `platforms` attribute, which is one of those 3 platform enums. So would we expand everything so that we have shell API metadata that matches these ad-hoc flags? That would end up exponentially growing and I'd worry about it being fragile. Since we already have those feature flags that are sorted by platform in shell-api, I'd say we should keep the generalized platform flags.

This is a slightly special case because we want to limit the creation of `Mongo` objects based on topology, _except for the initial mongo_ which gets created for all platforms.  So usually this check would go in the constructor of a class, and be automatically generated based on the attributes set by the topologies decorator. Since we haven't implemented this yet, it's not in the decorators, but that's how I expected it would go when we do write the platform-aware feature.

Reply:
I think is fine to keep it by platform, if that gets problematic for some reason we can figure something later.